### ISSUE TITLE
No difference between epi and found_epi (?)

### ISSUE THREAD

**Issue by PSmaruj** [jmschrei/ledidi#1]
Hello, I have tested the code provided in the "Usage" section. After getting the predicted sequence, I plotted `epi`, and `found_epi` to visually compare changes in CTCF binding between positions 487 and 537. The peak looks exactly the same. Moreover, I compared the input with the output sequence - there are 196 nucleotides changed, although the sequences are identical at interval 487-537. I expected my result to look exactly like the A part of the example graph. Does anybody have any suggestions on what could go wrong?

**Comment by jmschrei** [comment]
Howdy. Sorry that you're encountering difficulties. You'll need to provide more details, such as the model you're using and the exact code you used. It's difficult for me to diagnose the issue otherwise.

**Comment by PSmaruj** [comment]
Hello, the code I used can be found below.

```
#!/usr/bin/python

import numpy
import tensorflow as tf
import matplotlib.pyplot as plt

from ledidi import Ledidi
from ledidi import TensorFlowRegressor

from tensorflow.keras.models import load_model
from tensorflow.keras.utils import CustomObjectScope

# These two objects are necessary for the Basenji model

class GELU(tf.keras.layers.Layer):
    [def](url) __init__(self, **kwargs):
        super(GELU, self).__init__()

    def call(self, x, training):
        return tf.keras.activations.sigmoid(1.702 * x) * x

class StochasticShift(tf.keras.layers.Layer):
    def __init__(self, shift_max=0, pad='uniform', **kwargs):
        super(StochasticShift, self).__init__()

    def call(self, seq_1hot, training):
        return seq_1hot

custom_objects = {
    'StochasticShift': StochasticShift,
    'GELU': GELU
}

model = load_model("/home1/smaruj/ledidi/model_human.h5", custom_objects, compile=False)
model.compile()
regressor = TensorFlowRegressor(model=model)

# Index 687 is CTCF signal in GM12878
mask = numpy.zeros((1, 1024, 5313), dtype='float32')     
mask[:, :, 687] = 1.0                                     

mutator = Ledidi(regressor, mask=mask, l=1e2)

sequence = numpy.load("/home1/smaruj/ledidi/tutorials/CTCF-seqs.npz")['arr_0'].astype('float32')[0].reshape(1, 131072, 4)

epi = model.predict(sequence)

desired_epi = epi.copy()
desired_epi[0, 487:537, 687] = 0.0            #we want to this peak to disappear

edited_sequence = mutator.fit_transform(sequence, desired_epi) 
edited_sequence = tf.cast(edited_sequence, tf.float32)
edited_sequence = tf.expand_dims(numpy.array(edited_sequence), axis=0)

found_epi = model.predict(edited_sequence)[0, :, 687]

plt.plot(epi[0, 487:537, 687])
plt.plot(desired_epi[0, 487:537, 687])
plt.plot(found_epi[487:537])
```

**Comment by jmschrei** [comment]
Hey, so I finally got around to looking into this. I'm able to reproduce your issue where it's not getting rid of the peak. However, the training process looks totally different to me now than it did when I wrote the original paper. I don't know if this is an issue with TensorFlow or a new Basenji being released or anything. The number of edits made is way higher than I'm reporting. I even went back to the original code I used for the paper and made sure that I was using the correct hyperparameters and everything.

I messed around with the hyperparameters for a few hours trying to figure out what was going on and managed to get reasonable performance using a tau of 5. Although it did suggest making 1890 edits, which is way higher than when I did it last, the MSE was reduced from 182 when using the original sequence to 9.3 when using the edited sequence. I would suggest doing a hyperparameter search over tau, lr, and l and try to find a new good set of parameters. Unfortunately, I don't have time to debug the code right now and I'm not maintaining this package.
